platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/properties/PropertiesLoader.java

```diff
@@ -0,0 +1,226 @@
+ */
+package org.codice.ddf.platform.util.properties;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.function.BiFunction;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.text.StrSubstitutor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.io.support.PropertiesLoaderUtils;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Utility class that attempts several different methods for loading in properties files from the
+ * classpath or file system.
+ */
+public final class PropertiesLoader {
+    private static final Logger LOGGER = LoggerFactory.getLogger(PropertiesLoader.class);
+
+    private static final List<BiFunction<String, ClassLoader, Properties>>
+            PROPERTY_LOADING_STRATEGIES = ImmutableList.of(PropertiesLoader::attemptLoadWithSpring,
+            PropertiesLoader::attemptLoadWithSpringAndClassLoader,
+            PropertiesLoader::attemptLoadWithFileSystem,
+            PropertiesLoader::attemptLoadAsResource);
+
+    private PropertiesLoader() {
+        // static only!
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <K, V> Map<K, V> toMap(Properties properties) {
+        if (properties != null) {
+            final Set<Map.Entry<Object, Object>> entries = properties.entrySet();
+            Map<K, V> map = new HashMap<K, V>(entries.size() * 2);
+            for (Map.Entry<Object, Object> entry : entries) {
+                map.put((K) entry.getKey(), (V) entry.getValue());
+            }
+
+            return map;
+        }
+        return new HashMap<K, V>();
+    }
+
+    /**
+     * Load properties from a file with no classloader specified.
+     *
+     * @param propertiesFile
+     * @return Properties
+     */
+    public static Properties loadProperties(String propertiesFile) {
+        return loadProperties(propertiesFile, null);
+    }
+
+    /**
+     * Will attempt to load properties from a file using the given classloader. If that fails,
+     * several other methods will be tried until the properties file is located.
+     *
+     * @param propertiesFile
+     * @param classLoader
+     * @return Properties
+     */
+    public static Properties loadProperties(String propertiesFile, ClassLoader classLoader) {
+        Properties properties = new Properties();
+
+        if (propertiesFile == null) {
+            LOGGER.debug("Properties file must not be null.");
+            return properties;
+        }
+
+        Iterator<BiFunction<String, ClassLoader, Properties>> strategiesIterator =
+                PROPERTY_LOADING_STRATEGIES.iterator();
+
+        do {
```

Comment:
Unfortunately I don't think the streaming API will work in this instance. We need to cache the results of applying the function, but any variable used in a lambda should be final or effectively final, and `properties` is not. Any work-around AFAICT will result in redundant processing.